R/parameter_tuning.R

```diff
@@ -0,0 +1,218 @@
+create_parameter_grid = function(parameter_list, n_random_samples = NULL) {
+  #' Create a grid search of parameters and optionally take a random subset
+  #'
+  #' @param parameter_list A named list of parameters. Each element is either a list of values or a single value
+  #' @param n_random_samples How many samples to draw from the cross product list. If NULL, all elements will be taken
+  #'
+  #' @return A dataframe where each row contains one value from each parameter drawn from the list
+
+  # Create the cross product of all elements in the list. The column names will be taken from the
+  # names in the list
+  param_grid = expand.grid(parameter_list)
+
+  # Take a random sampling of rows if requested and the number is lower than the number of rows
+  if (!is.null(n_random_samples) && n_random_samples < nrow(param_grid)) {
+    param_grid = param_grid[sample(nrow(param_grid), n_random_samples), ]
+  }
+  return(param_grid)
+}
+
+
+compute_metric_per_trial = function(results_df, eval_function, params_df = NULL) {
+  #' Compute a scalar metric from each trial of an experiment.
+  #'
+  #' @param results_df A table of results, one row per trial
+  #' @param eval_function A function to evaluate each row, using the result and (optionally-null) parameters
+  #' @param params_df An optionally-null dataframe of paramters, one row per trial
+  #'
+  #' @return The list of metric values, one per row
+  #' The eval_function is run on the corresponding pairs of rows from results_df and params_df
+
+  # List of metrics from each run
+  metric_values = list()
+  # Iterate over all trials and compute the metric for each
+  for (i in seq_len(nrow(results_df))) {
+    # Get the shift predicted by the registration method
+    results = results_df[i, ]
+    # Get the map parameters defining the initial shift
+    if (is.null(params_df)) {
+      params = NULL
+    } else {
+      params = params_df[i, ]
+    }
+    # Compute a metric evaluating how well the predicted value matched the real one
+    metric = eval_function(results, params)
+    # Record this result
+    metric_values[[i]] = metric
```

Comment:
Suggest making `metric_values` a vector and not a list, and then indexing into it in the final line (46) with `metric_values[i]`. I think the only reason to make it a list is if there would ever be a case where each "trial" result is represented by more than one value. But as implemented here, each trial result ("metric") is a single value.